lib/ide/src/controller/module.rs

```diff
@@ -28,32 +56,161 @@ impl Location {
+        /// and update module state.
+        pub fn check_code_sync(&mut self, code:String) -> FallibleResult<()> {
+            let my_code = self.code();
+            if code != my_code {
+                self.logger.error(|| format!("The module controller ast was not synchronized with
+                    text editor content!\n >>> Module: {:?}\n >>> Editor: {:?}",my_code,code));
+                self.ast    = self.parser.parse(code,default())?;
+                self.id_map = default();
+            }
+            Ok(())
+        }
     }
 }
 
 impl Handle {
-    /// Receives a notification call when file with this module has been
-    /// modified by a third-party tool (like non-IDE text editor).
-    pub async fn file_externally_modified(&self) {
-        // TODO: notify underlying text/graph controllers about the changes
-        todo!()
+    /// Create a module controller for given location.
+    pub async fn new(location:Location, mut file_manager:fmc::Handle, mut parser:SharedParser)
+    -> FallibleResult<Self> {
+        let logger  = Logger::new(format!("Module Controller {}", location));
+        logger.info(|| "Loading module file");
+        let path    = location.to_path();
+        let content = file_manager.read(path).await?;
+        logger.info(|| "Parsing code");
+        let ast     = parser.parse(content,default())?;
+        logger.info(|| "Code parsed");
+        logger.trace(|| format!("The parsed ast is {:?}", ast));
+        let id_map  = default();
+        let data    = Controller {location,ast,file_manager,parser,id_map,logger};
+        Ok(Handle::new_from_data(data))
+    }
+
+    /// Save the module to file.
+    pub fn save_file(&self) -> impl Future<Output=Result<(),RpcError>> {
+        let (path,mut fm) = self.with_borrowed(|data| {
+            (data.location.to_path(),data.file_manager.clone_ref())
+        });
+        fm.write(path,self.code())
+        // TODO [ao] here save also the id_map and metadata.
+    }
+
+    #[cfg(test)]
+    fn new_mock
+    (location:Location, code:&str, id_map:IdMap, file_manager:fmc::Handle, mut parser:SharedParser)
+    -> FallibleResult<Self> {
+        let logger   = Logger::new("Mocked Module Controller");
+        let ast      = parser.parse(code.to_string(),id_map.clone())?;
+        let data     = Controller {location,ast,file_manager,parser,id_map,logger};
+        Ok(Handle::new_from_data(data))
+    }
+
+}
+
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use data::text::Span;
+    use data::text::Index;
+    use data::text::Size;
+    use json_rpc::test_util::transport::mock::MockTransport;
+    use uuid::Uuid;
+    use parser::SharedParser;
+    use basegl::display::shape::text::text_field::content::TextChange;
+    use basegl::display::shape::text::text_field::location::TextLocation;
+    use ast;
+    use wasm_bindgen_test::wasm_bindgen_test;
+    use ast::BlockLine;
+
+    #[wasm_bindgen_test]
```

Comment:
Is there a reason why we wouldn't want to test this natively as well?
Is it possible to have both wasm_bindgen_test and test attributes?

Reply:
because then you must manually run scala parser instance.